modules/tensor_mechanics/include/userobjects/CrystalPlasticitySlipRate.h

```diff
@@ -0,0 +1,70 @@
+/****************************************************************/
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*          All contents are licensed under LGPL V2.1           */
+/*             See LICENSE for full restrictions                */
+/****************************************************************/
+#ifndef CRYSTALPLASTICITYSLIPRATE_H
+#define CRYSTALPLASTICITYSLIPRATE_H
+
+#include "CrystalPlasticityUOBase.h"
+#include "RankTwoTensor.h"
+
+class CrystalPlasticitySlipRate;
+
+template<>
+InputParameters validParams<CrystalPlasticitySlipRate>();
+
+/**
+ * Crystal plasticity slip rate userobject class
+ * The virtual functions written below must be
+ * over-ridden in derived classes to provide actual values
+ */
+class CrystalPlasticitySlipRate : public CrystalPlasticityUOBase
+{
+ public:
+  CrystalPlasticitySlipRate(const InputParameters & parameters);
+
+  virtual void calcFlowDirection(unsigned int qp, std::vector<RankTwoTensor> & flow_direction) const = 0;
+  virtual bool calcSlipRate(unsigned qp, Real dt, std::vector<Real> & val) const = 0;
+  virtual bool calcSlipRateDerivative(unsigned qp, Real dt, std::vector<Real> & val) const = 0;
+
+ protected:
+  virtual void getSlipSystems();
+
+  virtual void readFileFlowRateParams();
+
+  virtual void getFlowRateParams();
+
+  bool _read_from_slip_sys_file;
```

Comment:
Check if you can make members constant that are only ever set in the initializer list (from input parameters for example).